File: web/room/net.js
// The client half of the protocol, which is deliberately three verbs long.
//
//   send(type, payload)   one intention. Accepted or refused, and told which.
//   poll()                one player's whole view of one frame.
//   presence(...)         a cursor. Allowed to be lost.
//
// There is no prediction here. A client that guessed at the result of its own
// command would be a second authority, and the entire experiment is about
// there being one -- so the ghost under the pointer is a picture, the command
// is a request, and what appears on the plot is whatever the host says
// appeared on the plot. At sixty ticks a second and a poll every 180 ms that
// is a fifth of a second of honesty, which is cheaper than a rollback.

export const state = {
  code: null,
  player: 0,
  catalogue: null,
  parts: null,
  view: null,       // the last frame the host sent
  err: null,
  onFrame: [],
  onRefusal: [],
  onHealth: [],
};

/// How the *connection* is doing, as opposed to how the room is doing.
///
/// These two used to be the same number and they are not the same question.
/// `view.sync` says whether this player's reconstruction of the room agrees
/// with the host's, which is the experiment. This says whether the screen is
/// being told about it at all, which is the difference between a factory that
/// stopped and a browser that stopped watching one -- and on screen those look
/// identical, which is exactly how a play session ends up reporting a freeze
/// that never happened.
export const health = {
  lag: 0,       // ms since a frame last arrived
  rtt: 0,       // how long the last poll took to answer
  misses: 0,    // polls in a row that did not answer
  behind: 0,    // ticks the room says this replica trails its own clock by
  live: false,  // whether the picture on screen is current
};

// ------------------------------------------------------------------- a seat
//
// A player id is a seat in one room, handed out by the host. It lived in this
// module and nowhere else, which meant a refresh -- or a click on a link in
// the left-hand menu -- threw it away and the next join took a *second* seat
// in a room where the first one still held everything that player had built.
// Two seats, one person, and a screen that no longer owned its own factory.
//
// So the browser keeps a token of its own. The token is not the identity the
// host uses -- it is the thing that proves which identity to hand back.

const SEAT = 'temporal-rooms/seat';
const LAST = 'temporal-rooms/room';

// Every access is guarded. `localStorage` throws rather than returns null in a
// private window and does not exist at all in the node test harness, and none
// of that is a reason for the game to fail to start.
function stored(k) {
  try { return localStorage.getItem(k); } catch { return null; }
}
function store(k, v) {
  try { localStorage.setItem(k, v); } catch { /* nothing to remember with */ }
}
function drop(k) {
  try { localStorage.removeItem(k); } catch { /* already gone */ }
}

let fallbackSeat = null;

/// This browser's token, minted once and kept.
export function seat() {
  const held = stored(SEAT);
  if (held) return held;
  const fresh = 'seat-' + Math.random().toString(36).slice(2, 10) + Date.now().toString(36);
  store(SEAT, fresh);
  // If there was nowhere to write it, at least be stable for this page: a
  // token that changes per request would be worse than not having one.
  if (stored(SEAT) !== fresh) return (fallbackSeat ||= fresh);
  return fresh;
}

/// The last room this browser was in, if it was in one.
export const lastRoom = () => stored(LAST);
const remember = code => store(LAST, code);
export const forget = () => drop(LAST);

// ------------------------------------------------------------------ the wire
//
// Every request answers `{ok}` or `{ok: false, error}` and never throws. It
// used to throw: `r.json()` on a connection the server dropped rejects, and
// nothing in `app.js` catches, so one refused socket killed whichever click
// handler was half-way through and left the screen holding a state the host
// had never agreed to. The poll loop below always had its own try/catch and so
// kept running, which is exactly why the failure looked like a freeze rather
// than an error -- the clock went on ticking and nothing else worked.

async function req(path, init, timeout = 0) {
  // A socket that is never going to answer must not be waited on forever. It
  // is not a hypothetical: the server reads a request head with a thirty
  // second timeout on it, and thirty seconds is long enough for a player to
  // conclude the game is broken, close the tab, and be right.
  const ctl = timeout && typeof AbortController === 'function' ? new AbortController() : null;
  const bell = ctl ? setTimeout(() => ctl.abort(), timeout) : null;
  try {
    const r = await fetch(path, ctl ? { ...init, signal: ctl.signal } : init);
    if (!r.ok) return { ok: false, error: `${path} answered ${r.status} ${r.statusText}` };
    return await r.json();
  } catch (e) {
    const gave_up = e && (e.name === 'AbortError' || e.name === 'TimeoutError');
    return {
      ok: false,
      error: gave_up
        ? `${path} did not answer within ${timeout} ms`
        : `${path} did not answer: ${(e && e.message) || e}`,
    };
  } finally {
    if (bell) clearTimeout(bell);
  }
}

const post = (path, body) => req(path, {
  method: 'POST',
  headers: { 'content-type': 'application/json' },
  body: JSON.stringify(body || {}),
});

const get = path => req(path);

// A failed answer is never cached: the catalogue is asked for once, and once
// used to mean once ever, including the once that failed.
export async function catalogue() {
  if (!state.catalogue) {
    const res = await get('/api/catalogue');
    if (!res.ok) return res;
    state.catalogue = res;
  }
  return state.catalogue;
}

export async function parts() {
  if (!state.parts) {
    const res = await get('/api/parts');
    if (!res.ok) return res;
    state.parts = res;
  }
  return state.parts;
}

export const goals = () => get('/api/goals');
export const rooms = () => get('/api/rooms');

function seated(res) {
  if (res.ok) {
    state.code = res.code;
    state.player = res.player;
    remember(res.code);
  }
  return res;
}

export async function host(name, seed, template) {
  return seated(await post('/api/host', { name, seed: seed || undefined, template, key: seat() }));
}

export async function join(code, name) {
  return seated(await post('/api/join', { code, name, key: seat() }));
}

/// Come back to the room this browser was last in.
///
/// The same route as a join, because the client cannot tell the difference: it
/// knows its token and the last code, and whether that is a seat or a stranger
/// is the host's question. `res.rejoined` is the host's answer. The name is
/// left off deliberately -- it was in the page that went away, and the room
/// still has it.
export async function rejoin() {
  const code = lastRoom();
  if (!code) return { ok: false, error: 'this browser has not been in a room' };
  // `back` says this is a reload and not a new player: without it a code left
  // in storage would take a fresh seat in whatever room still answers to it,
  // and a room reopened on the same seed answers to the same code.
  const res = seated(await post('/api/join', { code, key: seat(), back: true }));
  if (!res.ok) forget();
  return res;
}

/// Give up the seat this browser is holding, so the lobby is the lobby again.
export function leave() {
  stop();
  forget();
  state.code = null;
  state.player = 0;
  state.view = null;
}

/// One intention. The answer is small on purpose: everything the world looks
/// like afterwards arrives on the next poll, from the same reconstruction the
/// other player is being shown.
export async function send(type, payload) {
  const res = await post('/api/cmd', { code: state.code, player: state.player, type, payload });
  if (!res.ok) {
    state.err = res.error;
    for (const f of state.onRefusal) f(res.error, type);
  }
  return res;
}

/// Begin the clock. There is no matching stop.
export function begin() {
  return post('/api/start', { code: state.code });
}

export async function presence(cursor, selection, editing, view) {
  return post('/api/presence', {
    code: state.code, player: state.player, cursor, editing, view,
    // A selected wire is a string key rather than an installation id, and
    // presence is about buildings. Sending it as null is the honest answer.
    selection: typeof selection === 'number' ? selection : null,
  });
}

/// A design, built as a plant, for the 3D window. The document comes from the
/// room rather than from this browser: a client that could post any design
/// would be drawing something nobody else can see.
export function form(id, draft) {
  return post('/api/form', { code: state.code, id, draft: !!draft });
}

/// What every component in one machine is doing, at the phase the room's
/// clock puts its orbit in.
export function inside(id, draft) {
  return post('/api/inside', { code: state.code, id, draft: !!draft });
}

// ------------------------------------------------------------------ the poll
//
// A chain of `setTimeout`s, and it is worth being precise about what is now
// wrong with that and what is not.
//
// What is *not* wrong: the room falling behind. It does not any more. The
// server beats every room and every replica in it forward four times a second
// on a thread of its own, so a browser that stops asking costs the room
// nothing, and a browser that starts asking again gets a frame that is already
// current instead of triggering a minute of catch-up under a lock the other
// player is waiting on. That was the freeze, it was the *other* player's
// freeze, and it is fixed on the server.
//
// What is still wrong is this end of it. A browser is allowed to starve a
// timer: a background tab gets one a minute, a frozen page gets none until it
// is thawed, and a request on a dead socket can sit there until the server's
// own thirty-second read timeout gives up. In every one of those the room is
// fine and the picture is stale, which on screen is indistinguishable from the
// room being broken.
//
// So: no poll ever waits forever, no two are ever in flight at once, a second
// timer watches the first, coming back to the tab polls immediately rather
// than waiting out a throttled timeout, and how stale the picture is, is a
// number the screen can show rather than something a player has to guess at.

/// How long to wait for one frame before deciding this one is not coming.
/// Generous next to a 180 ms poll, because a frame is worth waiting for; short
/// next to thirty seconds, because a player is not.
const POLL_TIMEOUT = 5000;

/// How often the watchdog looks at the chain. Not a poll rate -- polls happen
/// on the chain -- just often enough that the number on screen moves while
/// nothing else is happening.
const WATCH = 1000;

const stamp = () => (typeof performance !== 'undefined' ? performance.now() : Date.now());

let timer = null;     // the chain
let watch = null;     // the watchdog over it
let period = 180;
let inflight = false;
let lastOk = 0;

/// One frame, or one reason there is not one.
///
/// Never two at once: a slow answer used to be followed by another request on
/// top of it, and a browser that fell behind would queue polls faster than the
/// server could answer them -- which is a way of making a stall permanent.
async function poll() {
  if (inflight || !state.code) return;
  inflight = true;
  const began = stamp();
  try {
    const v = await req(
      `/api/state?code=${state.code}&player=${state.player}`,
      undefined,
      POLL_TIMEOUT,
    );
    health.rtt = Math.round(stamp() - began);
    if (v.ok) {
      lastOk = stamp();
      health.misses = 0;
      health.behind = (v.sync && v.sync.behind) || 0;
      state.view = v;
      state.err = null;
      for (const f of state.onFrame) f(v);
    } else {
      health.misses++;
      state.err = v.error;
    }
  } catch (e) {
    // `req` does not throw, and a listener in `onFrame` might. A renderer that
    // fails must not stop the loop that would have drawn the next frame.
    health.misses++;
    state.err = String((e && e.message) || e);
  } finally {
    inflight = false;
    report();
  }
}

/// Say how stale the picture is, whether or not anything arrived.
function report() {
  health.lag = lastOk ? Math.round(stamp() - lastOk) : 0;
  health.live = !!lastOk && health.misses === 0 && health.lag < period * 6;
  for (const f of state.onHealth) f(health);
}

/// The tab came back, or the network did. Ask now.
function wake() {
  if (typeof document !== 'undefined' && document.visibilityState === 'hidden') return;
  poll();
}

function listen(on) {
  const verb = on ? 'addEventListener' : 'removeEventListener';
  if (typeof document !== 'undefined' && document[verb]) {
    document[verb]('visibilitychange', wake);
  }
  if (typeof window !== 'undefined' && window[verb]) {
    window[verb]('focus', wake);
    window[verb]('online', wake);
  }
}

export function start(p = 180) {
  stop();
  period = p;
  lastOk = 0;
  health.misses = 0;
  const chain = async () => {
    await poll();
    timer = setTimeout(chain, period);
  };
  chain();
  // The watchdog exists for the chain, not for the room: if the chain has been
  // throttled or a request is wedged, it asks again, and either way it keeps
  // the staleness on screen honest while nothing is arriving.
  watch = setInterval(() => {
    report();
    if (!inflight && stamp() - lastOk > period * 4) poll();
  }, WATCH);
  listen(true);
}

export function stop() {
  if (timer) clearTimeout(timer);
  if (watch) clearInterval(watch);
  timer = null;
  watch = null;
  listen(false);
}

export function onFrame(f) { state.onFrame.push(f); }
export function onRefusal(f) { state.onRefusal.push(f); }
export function onHealth(f) { state.onHealth.push(f); }

// ------------------------------------------------------------------ helpers

export function installs() {
  return (state.view && state.view.world.installs) || [];
}

export function byId(id) {
  return installs().find(i => i.id === id) || null;
}

export function proto(tag) {
  return (state.catalogue && state.catalogue.protos.find(p => p.tag === tag)) || null;
}

/// The plant's opinion of one installation, by name. The world document and
/// the simulation snapshot are joined here and nowhere else: names are the
/// only identity that survives a recompile, which is exactly why the
/// simulator carries state by them.
/// A wire has no `id` of its own -- it *is* the (from, to, item) triple that
/// made it -- so a selected one is spelled as a string and an installation
/// stays a number. Everything that consumes a selection can therefore go on
/// taking one argument.
export const wireKey = c => `w:${c.from}:${c.to}:${c.item}`;
export const wireOf = k =>
  typeof k === 'string' && k.startsWith('w:')
    ? (([, from, to, ...rest]) => ({ from: +from, to: +to, item: rest.join(':') }))(k.split(':'))
    : null;

export function plantOf(name) {
  const p = state.view && state.view.plant;
  if (!p || !p.classes) return null;
  return p.classes.find(c => c.name === name)
      || (p.storages || []).find(s => s.name === name)
      || (p.links || []).find(l => l.name === name)
      || null;
}

export function clock(t) {
  const s = Math.floor(t / 60);
  return `${Math.floor(s / 60)}:${String(s % 60).padStart(2, '0')}`;
}

export function num(n) {
  if (n === null || n === undefined) return '--';
  if (Math.abs(n) >= 1e6) return (n / 1e6).toFixed(1) + 'M';
  if (Math.abs(n) >= 1e4) return (n / 1e3).toFixed(1) + 'k';
  return Number.isInteger(n) ? String(n) : n.toFixed(1);
}

File: src/mp/net.rs
//! `room serve`: two browsers, one factory.
//!
//! ```text
//!   GET  /                      the game
//!   GET  /api/catalogue         what may be placed, and what it does
//!   GET  /api/parts             the components a machine is designed from
//!   GET  /api/kit               experiment 08's meshes, for the 3D view
//!   GET  /api/rooms             rooms that are open, for the lobby
//!   POST /api/host              open one, and get a code
//!   POST /api/join              join one, with the code
//!   POST /api/start             begin the clock. There is no matching stop.
//!   POST /api/cmd               one intention; accepted or refused
//!   POST /api/presence          a cursor. Lossy on purpose
//!   GET  /api/state             one player's whole view of one frame
//!   POST /api/form              a design, built as a plant, for the 3D view
//!   POST /api/inside            what every component in it is doing, and why
//!   GET  /api/goals             every template, for the developer panel
//! ```
//!
//! # What the server is, and is not
//!
//! It is a third small HTTP server in `std`, for the same reason the machine
//! designer got the second one: this experiment must be possible to throw
//! away. It binds to loopback, speaks HTTP/1.1 with `Connection: close`, and
//! holds the only thing in this project that is genuinely *stateful* -- the
//! rooms, their clocks, and one reconstruction per player.
//!
//! That statefulness is the difference from every server above it. Prototype
//! 0 and 1 could be stateless because `state(log, T)` is a pure function and
//! the browser could hold the log. A multiplayer room cannot: two browsers
//! cannot both hold the authority, and the whole question of the experiment is
//! what happens when they try.
//!
//! # Polling
//!
//! The client polls `/api/state` a few times a second and posts intentions as
//! they happen. No sockets, no server push, no frame synchronisation. That is
//! enough because nothing in the protocol is timing-sensitive: a command is
//! stamped by the host when it arrives, and a client that polls late gets a
//! later tick rather than a different history.

use super::room::Room;
use super::world::PlayerId;
use crate::http::{self, Req};
use crate::json::Json;
use crate::machine::design::Design;
use std::collections::HashMap;
use std::net::{TcpListener, TcpStream};
use std::sync::Mutex;
use std::time::Duration;

const ASSETS: &[(&str, &str, &str)] = &[
    ("/", "text/html; charset=utf-8", include_str!("../../web/room/index.html")),
    ("/room.css", "text/css; charset=utf-8", include_str!("../../web/room/room.css")),
    ("/app.js", "text/javascript; charset=utf-8", include_str!("../../web/room/app.js")),
    ("/net.js", "text/javascript; charset=utf-8", include_str!("../../web/room/net.js")),
    ("/world.js", "text/javascript; charset=utf-8", include_str!("../../web/room/world.js")),
    ("/bench.js", "text/javascript; charset=utf-8", include_str!("../../web/room/bench.js")),
    ("/panels.js", "text/javascript; charset=utf-8", include_str!("../../web/room/panels.js")),
    // Experiment 10's renderer, unchanged and unforked. The machine designer's
    // window is the machine designer's window; this prototype only decides
    // which design goes in it.
    // Served at the path it lives at, so that `../machine/form.js` resolves to
    // the same file whether it is being loaded by a browser over HTTP or by
    // the front end's own test harness off the disk.
    (
        "/machine/form.js",
        "text/javascript; charset=utf-8",
        include_str!("../../web/machine/form.js"),
    ),
];

static ROOMS: Mutex<Option<HashMap<String, Room>>> = Mutex::new(None);

fn with_rooms<R>(f: impl FnOnce(&mut HashMap<String, Room>) -> R) -> R {
    let mut g = ROOMS.lock().unwrap_or_else(|e| e.into_inner());
    f(g.get_or_insert_with(HashMap::new))
}

// -------------------------------------------------------------------- serve

pub fn serve(port: u16) -> std::io::Result<()> {
    let listener = bind(port)?;
    let addr = listener.local_addr()?;
    beat();
    println!("prototype 2 is at   http://{addr}/");
    println!("one browser hosts, the other joins with the code.");
    println!("ctrl-c to stop.");
    for stream in listener.incoming() {
        match stream {
            Ok(s) => {
                std::thread::spawn(move || {
                    if let Err(e) = handle(s) {
                        if !http::hung_up(e.kind()) {
                            eprintln!("request failed: {e}");
                        }
                    }
                });
            }
            Err(e) => eprintln!("accept failed: {e}"),
        }
    }
    Ok(())
}

/// The room's own thread.
///
/// Every room here used to be advanced by somebody's browser: `/api/state`
/// arrived, the room ran to the current tick, and the frame was cut from where
/// it landed. That works while every browser is polling and fails the moment
/// one is not -- and browsers stop polling constantly. A background tab is
/// throttled to a `setTimeout` a minute; a laptop that was shut sends nothing
/// at all. Every tick that passes meanwhile is a tick somebody's next poll has
/// to simulate in one call, holding [`ROOMS`] while it does, with the other
/// player's poll queued behind it. The person who froze was never the person
/// who walked away.
///
/// So the clock gets a thread. Four times a second it takes the lock, carries
/// every started room and every replica in it to the current tick, and puts
/// the lock down. The work is the same work -- the same ticks, the same
/// commands, the same hashes -- but it is spread across the beats it belongs
/// to rather than dropped on whichever request happens to arrive after a gap.
///
/// The sleep is at the top of the loop and unconditional, so a beat that runs
/// long cannot turn this into a thread that holds the lock forever: there is
/// always a quarter of a second in which requests get served.
fn beat() {
    std::thread::spawn(|| loop {
        std::thread::sleep(Duration::from_millis(super::HEARTBEAT_MS));
        with_rooms(|rs| {
            for r in rs.values_mut() {
                r.heartbeat();
            }
        });
    });
}

fn bind(port: u16) -> std::io::Result<TcpListener> {
    let mut last = None;
    for p in port..port + 8 {
        match TcpListener::bind(("127.0.0.1", p)) {
            Ok(l) => return Ok(l),
            Err(e) => last = Some(e),
        }
    }
    Err(last.unwrap())
}

fn handle(stream: TcpStream) -> std::io::Result<()> {
    let Some(req) = http::accept(&stream)? else { return Ok(()) };
    let (status, mime, payload) = route(&req);
    http::reply(&stream, status, mime, &payload)
}

const MIME: &str = "application/json; charset=utf-8";

fn route(req: &Req) -> (&'static str, &'static str, String) {
    if req.method == "GET" {
        if let Some((_, m, body)) = ASSETS.iter().find(|(p, _, _)| *p == req.path) {
            return ("200 OK", m, body.to_string());
        }
        match req.path.as_str() {
            "/api/catalogue" => return ok(super::kit::catalogue()),
            "/api/parts" => {
                return ok(Json::obj()
                    .set("ok", true)
                    .set("parts", Design::catalogue())
                    .set("portKinds", crate::machine::design::port_kinds())
                    .set("substances", crate::machine::design::substances()))
            }
            "/api/kit" => return ok(crate::machine::form::kit_json()),
            "/api/goals" => return ok(goals()),
            "/api/rooms" => return ok(rooms()),
            "/api/state" => {
                let (code, player) = (req.q("code"), req.q("player"));
                return answer(with_rooms(|rs| match rs.get_mut(&code) {
                    Some(r) => r.view(player.parse().unwrap_or(0)),
                    None => Err(format!("there is no room {code}")),
                }));
            }
            _ => {}
        }
    }
    if req.method == "POST" {
        match req.path.as_str() {
            "/api/host" => return answer(host(&req.json())),
            "/api/join" => return answer(join(&req.json())),
            "/api/start" => return answer(start(&req.json())),
            "/api/cmd" => return answer(command(&req.json())),
            "/api/presence" => return answer(presence(&req.json())),
            "/api/form" => return answer(form(req)),
            "/api/inside" => return answer(inside(req)),
            _ => {}
        }
    }
    ("404 Not Found", MIME, err("no such route").to_string())
}

fn ok(j: Json) -> (&'static str, &'static str, String) {
    ("200 OK", MIME, j.to_string())
}

fn answer(r: Result<Json, String>) -> (&'static str, &'static str, String) {
    match r {
        Ok(j) => ("200 OK", MIME, j.to_string()),
        Err(e) => ("200 OK", MIME, err(&e).to_string()),
    }
}

fn err(msg: &str) -> Json {
    Json::obj().set("ok", false).set("error", msg)
}

// ------------------------------------------------------------------- routes

/// Open a room. The seed decides the code and the goal together, so a room
/// somebody wants to play again is one number long.
fn host(j: &Json) -> Result<Json, String> {
    let seed = j.at("seed").as_u64().unwrap_or_else(fresh_seed);
    let template = j.at("template").as_str().filter(|t| !t.is_empty());
    let name = j.at("name").as_str().unwrap_or("host").to_string();
    let key = j.at("key").as_str().unwrap_or_default().to_string();
    with_rooms(|rs| {
        let mut room = Room::open(seed, template);
        if rs.contains_key(&room.code) {
            return Err(format!("room {} is already open", room.code));
        }
        let code = room.code.clone();
        // The host is a player like any other: their browser is a client of
        // the authority, even though the authority is in the same process.
        //
        // The clock is *not* started here. Section 19 of the brief asks for
        // the objective to be on screen before anybody builds, and that is the
        // only pause the game has: once `/api/start` is called there is no
        // matching stop.
        let (id, _) = room.join_as(&name, &key)?;
        let goal = room.goal.clone();
        let progress = room.host.progress();
        rs.insert(code.clone(), room);
        Ok(Json::obj()
            .set("ok", true)
            .set("code", code)
            .set("player", id as i64)
            .set("seed", Json::big(seed as u128))
            .set("goal", goal.to_json(&progress)))
    })
}

/// Arrive, or come back.
///
/// The same route for both, because the client cannot tell which it is doing:
/// a browser that was refreshed knows only its own token and the last code it
/// was in. Whether that is a seat or a stranger is the room's question, and
/// the answer comes back as `rejoined` so the screen can say so.
///
/// An absent name means "whatever I was called before", which is what a
/// reload has to mean -- the name was in the page that went away.
///
/// `back` is a client saying it is only interested in a seat it already has.
/// Without it a stale code in a browser's storage would take a *new* seat in
/// whatever room happens to answer to it -- and a code is derived from a seed,
/// so a room reopened on the same seed answers to the same code. That is a
/// phantom player in somebody else's room, arriving from a tab nobody opened.
fn join(j: &Json) -> Result<Json, String> {
    let code = j.at("code").as_str().unwrap_or_default().to_uppercase();
    let name = j.at("name").as_str().unwrap_or_default().to_string();
    let key = j.at("key").as_str().unwrap_or_default().to_string();
    let back = j.at("back").as_bool().unwrap_or(false);
    with_rooms(|rs| {
        let room = rs.get_mut(&code).ok_or(format!("there is no room {code}"))?;
        if back && !room.seated(&key) {
            return Err(format!("your seat in {code} is not there any more"));
        }
        let (id, rejoined) = room.join_as(&name, &key)?;
        Ok(Json::obj()
            .set("ok", true)
            .set("code", code.clone())
            .set("player", id as i64)
            .set("rejoined", rejoined)
            // Seat one opened the room, and seat one is the only seat with a
            // start button on it. A host that reloaded before starting the
            // clock has to get that button back, so the answer says which seat
            // this is rather than leaving the browser to remember.
            .set("host", id == 1)
            .set("name", room.player(id).map(|p| Json::Str(p.name.clone())))
            .set("joinedAt", room.player(id).map(|p| Json::Int(p.joined as i128))))
    })
}

fn start(j: &Json) -> Result<Json, String> {
    let code = j.at("code").as_str().unwrap_or_default().to_uppercase();
    with_rooms(|rs| {
        let room = rs.get_mut(&code).ok_or(format!("there is no room {code}"))?;
        room.start();
        Ok(Json::obj().set("ok", true).set("tick", room.now()))
    })
}

/// One intention.
///
/// The answer is deliberately small: whether it was accepted, and the
/// canonical `(tick, sequence)` if it was. Everything else the client needs
/// arrives on the next poll, from the same reconstruction every other client
/// is being shown.
fn command(j: &Json) -> Result<Json, String> {
    let code = j.at("code").as_str().unwrap_or_default().to_uppercase();
    let player: PlayerId = j.at("player").as_u64().unwrap_or(0) as PlayerId;
    let act = super::cmd::Cmd::from_json(
        &Json::obj()
            .set("type", j.at("type").clone())
            .set("payload", j.at("payload").clone()),
    )
    .map(|c| c.act)?;
    with_rooms(|rs| {
        let room = rs.get_mut(&code).ok_or(format!("there is no room {code}"))?;
        match room.submit(player, act) {
            Ok(c) => Ok(Json::obj().set("ok", true).set("command", c.to_json())),
            // A refusal is an answer, not an error: the client is told which
            // command and why, and the room carries on exactly as it was.
            Err(e) => Ok(Json::obj().set("ok", false).set("refused", true).set("error", e)),
        }
    })
}

/// A cursor, a selection, and which window somebody is looking at.
///
/// None of it is ordered, none of it is hashed, and losing a packet of it
/// costs nothing at all.
fn presence(j: &Json) -> Result<Json, String> {
    let code = j.at("code").as_str().unwrap_or_default().to_uppercase();
    let id: PlayerId = j.at("player").as_u64().unwrap_or(0) as PlayerId;
    with_rooms(|rs| {
        let room = rs.get_mut(&code).ok_or(format!("there is no room {code}"))?;
        let Some(p) = room.players.iter_mut().find(|p| p.id == id) else {
            return Err("you are not in this room".into());
        };
        let c = j.at("cursor");
        p.cursor = match (c.at("x").as_f64(), c.at("y").as_f64()) {
            (Some(x), Some(y)) => Some((x, y)),
            _ => None,
        };
        p.selection = j.at("selection").as_u64();
        p.editing = j.at("editing").as_u64();
        if let Some(v) = j.at("view").as_str() {
            p.view = v.to_string();
        }
        Ok(Json::obj().set("ok", true))
    })
}

/// A design, built as a plant, for the 3D window.
///
/// The document comes from the room rather than from the browser: a client
/// that could post any design here would be drawing something nobody else can
/// see. `draft` asks for the one being edited; without it you get the design
/// that is actually running.
///
/// The style is not a choice here, the way it is in the designer. The room's
/// 3D window is an *editor*: the only reason to open a machine is to look at
/// its components and place another one, and `works` and `hall` both answer
/// that by putting walls and a roof between the player and the thing they
/// came in to work on. So a room always builds the yard -- a slab, and the
/// plant standing on it in the open -- and the enclosure a machine would
/// otherwise wear is a question for the designer, where nobody is editing.
fn form(req: &Req) -> Result<Json, String> {
    let j = req.json();
    let code = j.at("code").as_str().unwrap_or_default().to_uppercase();
    let id = j.at("id").as_u64().unwrap_or(0);
    let want_draft = j.at("draft").as_bool().unwrap_or(false);
    let grade = req.q("grade");
    let seed = req.q("seed").parse().unwrap_or(0);
    let design = with_rooms(|rs| {
        let room = rs.get(&code).ok_or(format!("there is no room {code}"))?;
        let i = room.host.world.get(id).ok_or("there is no such machine")?;
        let d = if want_draft { i.draft.clone().or_else(|| i.design.clone()) } else { i.design.clone() };
        d.ok_or_else(|| "that installation has no design".to_string())
    })?;
    let ask = crate::machine::form::Ask {
        style: crate::machine::form::Style::Yard,
        world: seed,
        grade: crate::machine::form::Grade::by_tag(&grade).unwrap_or_default(),
    };
    let scene = crate::machine::form::build(&design, ask)?;
    Ok(scene
        .to_json()
        .set("ok", true)
        .set("design", design.to_json())
        .set("source", design.emit()))
}

/// What every component inside one machine is doing, and why it is not doing
/// more.
///
/// The phase is the room's clock read through the machine's own orbit: a
/// designer tick is a game second, and a settled machine at second 4,000 is
/// indistinguishable from itself at `transient + 4000 % period`. So the
/// diagnostics are the machine's, at the moment the room is at, without the
/// inner simulation ever being stepped by the outer one.
///
/// What this deliberately does **not** show is the machine starved by the
/// *world*. Inside, it is running on its design's own supply; outside, the
/// world's inspector says whether the bay it draws from is empty. Two
/// altitudes, two answers, and the outer one is the authority on whether the
/// machine turned at all.
fn inside(req: &Req) -> Result<Json, String> {
    let j = req.json();
    let code = j.at("code").as_str().unwrap_or_default().to_uppercase();
    let id = j.at("id").as_u64().unwrap_or(0);
    let want_draft = j.at("draft").as_bool().unwrap_or(false);
    let (design, now) = with_rooms(|rs| {
        let room = rs.get(&code).ok_or(format!("there is no room {code}"))?;
        let i = room.host.world.get(id).ok_or("there is no such machine")?;
        let d = if want_draft {
            i.draft.clone().or_else(|| i.design.clone())
        } else {
            i.design.clone()
        };
        Ok::<_, String>((d.ok_or_else(|| "that installation has no design".to_string())?, room.now()))
    })?;
    let c = crate::machine::orbit::compile(&design)?;
    let r = crate::machine::eval::report(&design, &c);
    let phase = c.equivalent_tick(now / super::DESIGN_TICK);
    let m = c.state_at(&design, phase)?;
    Ok(crate::machine::snap::render(&design, &m, &r)
        .set("ok", true)
        .set("phase", phase as i64)
        .set("period", c.period as i64)
        .set("transient", c.transient as i64))
}

fn rooms() -> Json {
    with_rooms(|rs| {
        let mut open: Vec<Json> = rs
            .values()
            .map(|r| {
                Json::obj()
                    .set("code", r.code.clone())
                    .set("goal", r.goal.title.clone())
                    .set("brief", r.goal.brief())
                    .set("players", r.players.len() as i64)
                    .set("tick", r.now())
                    .set("seconds", super::as_secs(r.now()))
            })
            .collect();
        open.sort_by_key(|j| j.at("code").as_str().unwrap_or_default().to_string());
        Json::obj().set("ok", true).set("rooms", Json::Arr(open))
    })
}

/// Every template, with the range each of its numbers is drawn from -- for a
/// developer who wants to force one rather than roll for it.
fn goals() -> Json {
    Json::obj().set("ok", true).set(
        "templates",
        Json::Arr(
            super::goal::TEMPLATES
                .iter()
                .map(|t| {
                    let sample = super::goal::Goal::of_seed(1, Some(t.id));
                    Json::obj()
                        .set("id", t.id)
                        .set("family", t.family.word())
                        .set("title", t.title)
                        .set("note", t.note)
                        .set("example", sample.brief())
                })
                .collect(),
        ),
    )
}

/// A seed nobody chose. The clock is the only entropy this program has, and
/// the only place it is allowed to be used: everything downstream of the seed
/// is a pure function of it.
fn fresh_seed() -> u64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    let t = SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default();
    super::hash64(&t.as_nanos().to_le_bytes())
}
